File: src/main/java/org/tp23/xgen/XGenNodeList.java
/*
 * Copyright teknopaul 2014 LGPL
 */
package org.tp23.xgen;

import javax.xml.xpath.XPathExpressionException;

import org.w3c.dom.Attr;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

/**
 * This NodeList implementation is a more convenient than the org.w3c one. It support mutating the elements in the list
 * and it  can be iterated with foreach loops.
 * 
 * This can result in jQuery style coding where lists of results are treated the same as single results. 
 * 
 * @author teknopaul
 *
 */
public class XGenNodeList extends MutableNodelList {
	
	private XGen xGen;
	
	public XGenNodeList(XGen xGen) {
		this.xGen = xGen;
	}

	public XGenNodeList(XGen xGen, NodeList nodeList) {
		super(nodeList);
		this.xGen = xGen;
	}
	
	public static XGenNodeList createSingleNodeList(XGen xGen, Node node) {
		XGenNodeList xGenNodeList = new XGenNodeList(xGen);
		xGenNodeList.add(node);
		return xGenNodeList;
	}
	
	public XGenNodeList setTextContent(String string) {
		for (Node node : this) {
			node.setTextContent(string);
		}
		return this;
	}
	public XGenNodeList setTextContent(String...strings) {
		for (int i = 0 ; i < getLength() && i < strings.length; i++) {
			item(i).setTextContent(strings[i]);
		}
		return this;
	}
	
	public XGenNodeList setAttribute(String name, String value) {
		for (Node node : this) {
			if (node instanceof Element) {
				((Element)node).setAttribute(name, value);
			}
		}
		return this;
	}
	public XGenNodeList setAttribute(Attr... newAttr) {
		for (int i = 0 ; i < getLength() && i < newAttr.length; i++) {
			Node node = item(i);
			if (node instanceof Element) {
				((Element)node).setAttributeNode(newAttr[i]);
			}
		}
		return this;
	}
	
	public XGenNodeList appendChild(Node newChild) {
		final XGenNodeList tailNodes = new XGenNodeList(xGen);
		for (Node node : this) {
			Node clone = newChild.cloneNode(true);
			node.appendChild(clone);
			tailNodes.add(clone);
		}
		return tailNodes;
	}
	public XGenNodeList appendChild(Node... newChild) {
		final XGenNodeList tailNodes = new XGenNodeList(xGen);
		for (int i = 0 ; i < getLength() && i < newChild.length; i++) {
			item(i).appendChild(newChild[i]);
			tailNodes.add(newChild[i]);
		}
		return tailNodes;
	}
	
	public XGenNodeList each(NodeMutator mutator) {
		final XGenNodeList tailNodes = new XGenNodeList(xGen);
		for (Node node : this) {
			tailNodes.add(mutator.each(node));
		}
		return tailNodes;
	}
	
	public XGenNodeList create(String xGenPath) throws XGenExpressionException {
		return xGen.create(this, xGenPath);
	}
	
	public XGenNodeList select(String xPath) throws XPathExpressionException {
		XGenNodeList xGenNodeList = new XGenNodeList(xGen);
		for (Node node : this) {
			xGenNodeList.addElements(xGen.select((Element)node, xPath));
		}
		return xGenNodeList;
	}
}

File: src/main/js/xgen-nodelist.js

// To make xGenNodeList[i] work, use this syntax
// Array.prototype.push.call(this, item);

var XGenNodeList = function(xGen, nodeList){
	this.xGen = xGen;
	this.list = [];
	if (nodeList) this.addElements(nodeList);
};

// @static
XGenNodeList.createSingleNodeList = function(xGen, node) {
	var xGenNodeList = new XGenNodeList(xGen);
	xGenNodeList.add(node);
	return xGenNodeList;
};

XGenNodeList.prototype.item = function(index) {
	return this.list[index];
};

Object.defineProperty(XGenNodeList.prototype, "length", {
	get: function() {
		return this.list.length;
	},
	set: function(len) {
		// noop
	}
});

XGenNodeList.prototype.getLength = function() {
	return this.list.length;
};

XGenNodeList.prototype.add = function(node) {
	this.list.push(node);
};

XGenNodeList.prototype.addElements = function(nodeList) {
	// Dom nodeList or XGenNodeList
	if (nodeList.item) {
		for (var i = 0 ; i < nodeList.length ; i++) {
			this.list.push(nodeList.item(i));
		}
	}
	// xPath array (TODO fix this so XPath returns proper NodeLists)
	else if (nodeList.length) {
		for (var i = 0 ; i < nodeList.length ; i++) {
			this.list.push(nodeList[i]);
		}
	}
};

XGenNodeList.prototype.setTextContent = function(string) {
	for (var iter = this.iterator(); iter.hasNext() ; ) {
		var node = iter.next();
		var text = this.xGen.document.createTextNode(string);
		node.appendChild(text);
	}
	return this;
};
XGenNodeList.prototype.setAttribute = function(name, value) {
	for (var iter = this.iterator(); iter.hasNext() ; ) {
		var node = iter.next();
		//TODO if (node instanceof Element) {
			node.setAttribute(name, value);
		//}
	}
	return this;
}
XGenNodeList.prototype.appendChild = function(newChild) {
	var tailNodes = new XGenNodeList(this.xGen);
	for (var iter = this.iterator(); iter.hasNext() ; ) {
		var node = iter.next();
		var clone = newChild.cloneNode(true);
		node.appendChild(clone);
		tailNodes.add(clone);
	}
	return tailNodes;
};

XGenNodeList.prototype.create = function(xGenPath) /*throws XGenExpressionException*/ {
	return this.xGen.create(this, xGenPath);
};
XGenNodeList.prototype.select(xPath) throws XPathExpressionException {
	var xGenNodeList = new XGenNodeList(xGen);
	for (var iter = this.iterator(); iter.hasNext() ; ) {
		var node = iter.next();
		xGenNodeList.addElements(xGen.select(node, xPath));
	}
	return xGenNodeList;
};

// TODO this should be better in JS 1.7, does not seem to work in node, investigate
XGenNodeList.prototype.iterator = function() {
	var iter = {
		idx : 0,
		list : this.list,
		hasNext : function() {
			return this.idx < this.list.length;
		},
		next : function() {
			if (this.idx >= this.list.length) {
				throw new StopIteration();
			}
			return this.list[this.idx++];
		}
	}
	return iter;
};

if (exports) {
	exports.XGenNodeList = XGenNodeList;
}